### PR TITLE
Added Trubrics Feedback Collector to DemoGPT

### PR DESCRIPTION
**Pull request by PBozmarov**

### CODE REVIEW

**Comment by PBozmarov** [comment]
Hey @melih-unsal, this PR enables you to gather and analyze feedback for the results and sources of the DemoGPT model by integrating the Trubrics SDK. You are then also able to analyse the feedback received through registration on the Trubrics platform.

## Setup
To store feedback in your Trubrics account, follow these steps:

1. Register for a free account with Trubrics [here](https://trubrics.streamlit.app/)
2. Input your Trubrics email & password into [st.secrets](https://docs.streamlit.io/streamlit-community-cloud/get-started/deploy-an-app/connect-to-data-sources/secrets-management)

## Preview

Here is an example of gathering and storing feedback using the faces feature for the generated code from DemoGPT.
Screenshot 2023-07-24 at 16 42 19

We have also integrated feedback into the Streamlit app with the generated model:
Screenshot 2023-07-24 at 16 33 56


In the following image, we can see the feedback saved in our Trubrics account.
Screenshot 2023-07-24 at 16 41 42




## Useful links
1. See [here](https://trubrics.github.io/trubrics-sdk/integrations/streamlit/) for information on how to integrate Trubrics with Streamlit.
2. [Here](https://trubrics.github.io/trubrics-sdk/trubrics_platform/feedback_components/) you can find descriptions of the various feedback components (thumbs, faces, custom text).

We believe it would be fascinating to investigate the possibility of the feedback collector being automatically incorporated into the code when the model is generated, as opposed to being added externally as we did. Please let me know what you think.

If you have any questions, feel free to contact me.

File: .gitignore [hunk 1]
```diff
@@ -23,7 +25,8 @@ src/docs_explanation/*
 src/docs_explanation_some/*
 src/docs_summary/*
 
-
+# environment
+demo_gpt/*
```
Comment (jeffkayne) [inline .gitignore]: This doesn't look like its used by the maintainer, so its best to keep it out of the PR!

File: src/prompt_based/app.py [hunk 2]
```diff
@@ -33,13 +34,25 @@ def generate_response(txt):
...
 
 
+if "success" not in st.session_state:
+    st.session_state["success"] = False
+
+if "streamlit_code" not in st.session_state:
+    st.session_state["success"] = ""
+
+if "example_submitted" not in st.session_state:
+    st.session_state["example_submitted"] = True
```
Comment (jeffkayne) [inline src/prompt_based/app.py]: Is this used?

File: src/prompt_based/app.py [hunk 3]
```diff
@@ -76,58 +89,82 @@ def generate_response(txt):
...
+            if st.session_state["success"]:
+                break
+        else:
+            bar.progress(100, text=PROGRESS_BAR_TEXTS["failed"])
+
+if st.session_state["success"]:
+    with st.expander("Code"):
+        st.code(st.session_state["streamlit_code"])
+
+    model = "OpenAI"
```
Comment (jeffkayne) [inline src/prompt_based/app.py]: Can we read this in programatically instead of setting a variable?

File: src/prompt_based/model.py [hunk 4]
```diff
@@ -62,7 +62,11 @@ def normalize(self,code):
         Returns:
             str: refined version of the code
         """
-        return unicodedata.normalize('NFKD', code).encode('ascii', 'ignore').decode('utf-8')
+        return (
+            unicodedata.normalize("NFKD", code)
+            .encode("ascii", "ignore")
+            .decode("utf-8")
+        )
```
Comment (jeffkayne) [inline src/prompt_based/model.py]: To make reviews easier for the maintainer, always try to keep their formatting, especially if no code has changed. Prevents them spending time looking for differences where there are none

File: src/prompt_based/model.py [hunk 5]
```diff
@@ -328,4 +356,5 @@ def __call__(self, topic, title, code, test_code, progress_func, success_func):
         refined_code = self.refine_code(streamlit_code)
         progress_func(100, "Redirecting to the demo page...")
         success_func()
+
```
Comment (jeffkayne) [inline src/prompt_based/model.py]: same, can remove lines like this to keep PR nice and clean :)

File: src/prompt_based/model.py [hunk 6]
```diff
@@ -52,17 +52,17 @@ def refine_code(self, code):
...
+                code (str): The code to be refined.
 
-        Returns:
-            str: refined version of the code
-        """
-        return unicodedata.normalize('NFKD', code).encode('ascii', 'ignore').decode('utf-8')
+            Returns:
+                str: refined version of the code
+            """
+            return unicodedata.normalize('NFKD', code).encode('ascii', 'ignore').decode('utf-8')
```
Comment (jeffkayne) [inline src/prompt_based/model.py]: Remove these indents

**Pull request merged**